### PR TITLE
use rust-overlay for rust compiler+src, and cargo for crates

### LINKED ISSUE #13: vs code reports: Rust Tools missing
After install of rustycode plugin, status line reports: "Rust Tools Missing"
Seems, the binary `rustsym` is missing

### LINKED ISSUE #35: overlay which mimic rustup behaviour
See: https://www.mail-archive.com/nix-dev@lists.science.uu.nl/msg33296.html

From:
https://www.reddit.com/r/rust/comments/5zj69r/nixnixos_users/

### LINKED ISSUE #36: how to update bsys env to rustc-1.16 and co
where do I find the fix to rustc-1.15.1 and how to update to 1.16.1? Would it be easier to work with overlays as mentioned in #35 ?

### LINKED ISSUE #63: rustfmt 0.7 VS recent rustfmt 0.83
rustfmt on workstation ist quite 'old' .... the recent rustfmt version is 0.83 (on my system)

The results from running these different versions produce different output.

So, I think we need a fast update solution, otherwise we have problems running the ci tests on different systems. Of course I could downgrade, but this only delays the next update problems.

### PR DESCRIPTION
**Pull request by steveeJ**
Fixes #36 
Fixes #35 
Fixes #13 
Fixes #63 

TODO
- [x] Figure out CI failures

### CODE REVIEW

**Comment by mmaechtel** [comment]
labshell exit and again labshell bsys:
```
+ NIX_SHELL_DRV=/home/maechtel/nix-expressions/scripts/../
+ NIX_SHELL_DRVATTR=shell_bsys
+ exec nix-shell --pure --argstr shDrv /home/maechtel/nix-expressions/scripts/../ -A shell_bsys /home/maechtel/nix-expressions/scripts/../
downloading ‘https://static.rust-lang.org/dist/channel-rust-stable.toml’... [0/0 KiB, 0.0 KiB/s]
    Updating registry `https://github.com/rust-lang/crates.io-index`
  Installing racer v2.0.6
error: binary `racer` already exists in destination as part of `racer v2.0.6`
Add --force to overwrite
    Updating registry `https://github.com/rust-lang/crates.io-index`
  Installing rustfmt v0.8.3
error: binary `cargo-fmt` already exists in destination as part of `rustfmt v0.8.3`
binary `rustfmt` already exists in destination as part of `rustfmt v0.8.3`
Add --force to overwrite
    Updating registry `https://github.com/rust-lang/crates.io-index`
  Installing rustsym v0.3.1
error: binary `rustsym` already exists in destination as part of `rustsym v0.3.1`
Add --force to overwrite
maechtel@ct-bsys-ss17-0 ✓ »shell_bsys@1« ~/nix-expressions
```

seems it tries to install again, even if already binaries are installed. Should only install new ones ...

**Comment by steveeJ** [comment]
> seems it tries to install again, even if already binaries are installed. Should only install new ones ...

I have noticed that too. I haven't found an elegant way to test for installed packages in cargo yet.

**Comment by mmaechtel** [comment]
Maybe some kind of wrapper:
- knows, that when labshell bsys runs, that there is no new version of the overlay and doesn't call cargo install
- when it sees a new overlay version it runs cargo install with --force, to install new versions

**Comment by steveeJ** [comment]
> Maybe some kind of wrapper:
> 
>  * when labshell bsys runs but there is no new version of the overlay, it doesn't call cargo install
>  * when it sees a new overlay version it runs cargo install with --force, to install new versions

yet another wrapper 🙈 I'll give it some thought, but I'll also look for a native nix solution instead of using cargo.

File: shells/default.nix [hunk 1]
```diff
@@ -84,7 +82,20 @@ let
...
     rust = ''
-      export RUST_SRC_PATH="${pkgs.rustStable.rustc.src}/src"
+      export RUST_SRC_PATH="${pkgs.rustChannels.stable.rust-src}/lib/rustlib/src/rust/src/"
+
+      export CARGO_INSTALL_ROOT=/var/tmp/cargo
+      mkdir -p $CARGO_INSTALL_ROOT
+      export PATH=$CARGO_INSTALL_ROOT/bin:$PATH
+
+      for crate in racer rustfmt rustsym; do
+        cargo install --list | grep $crate 2>&1 1>/dev/null
```
Comment (steveeJ) [inline shells/default.nix]: @mmaechtel this doesn't install duplicates. We're still missing an update mechanism for crates. One step at a time though!?

**Comment by steveeJ** [comment]
@mmaechtel PR updated with logic to install crates with specific versions and installed version detection.

**Pull request merged**